### PR TITLE
Readme describing tuple representations in Greenplum

### PR DESCRIPTION
**Pull request by asimrp**
The content is copied from an internal document, barring formatting changes.
Thank you Xin for digging out that document.  Let's bring the content close to
source code so that it's alive and useful.

### CODE REVIEW

**Comment by asimrp** [comment]
@kavinderd can executor experts like you please verify that the content is valid?

File: src/backend/executor/README.tuples [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+attribute involves iterating through the first M-1 attributes,
+calculating their actual length (fixed or variable, null or not), the
+adding to an offset.
+
+MemTuple is an optimized representation which speeds-up non-sequential
+access of the At query start time, for each slot (or each place we use
+MemTuple), we compute a binding data structure from the schema
+(TupleDesc).  For each attribute in the tuple, the binding structure
+contains information that determines where the attribute is located in
+the MemTuple.  It consists the following:
```
Comment (kavinderd) [inline src/backend/executor/README.tuples]: It consists *of* the following

File: src/backend/executor/README.tuples [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+ 1. The physical attribute number.  Physical attr number may differ
```
Comment (kavinderd) [inline src/backend/executor/README.tuples]: Physical attr *numbers*

File: src/backend/executor/README.tuples [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+Tuples are passed around in generic TupleTableSlot objects. Interested
+parties need to be able to decode the data and see the values. The
+current tuple representation provides information on how to decode the
+data into typed attributes
+
+Why do we need more than one tuple representations?
+---------------------------------------------------
+
+Data comes from disk in a disk-oriented representation: serialized in
+a buffer, and including MVCC columns for heap tables.  That is
```
Comment (kavinderd) [inline src/backend/executor/README.tuples]: and *includes*

File: src/backend/executor/README.tuples [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+sub-optimal for fast processing.  Execution engine does not usually
```
Comment (kavinderd) [inline src/backend/executor/README.tuples]: *The* execution engine...

File: src/backend/executor/README.tuples [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+   is_null) arrays.
+
+Tuple Representations Stored in a TupleTableSlot
+------------------------------------------------
+
+HeapTuple
+
+This is identical to the on-disk representation in a heap table. It's
+a serialized (or "materialized") tuple, stored either in a buffer that
+belongs to the buffer manager, or palloc’ed.  Resource management cares
```
Comment (kavinderd) [inline src/backend/executor/README.tuples]: or *palloc allocated onto the process' heap* (I think this makes it more clear)

File: src/backend/executor/README.tuples [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+about which case, as it needs to hold pins etc in the first case.
```
Comment (kavinderd) [inline src/backend/executor/README.tuples]: I would rewrite the last sentence to:

'Where a heap tuple is stored is important for resource management as tuples in the buffer need to be correctly pinned'

File: src/backend/executor/README.tuples [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+responsibility of the generating plan node to be sure these resources
+are not released for as long as the virtual tuple needs to be valid.
+
+Tuples that need to be copied anywhere (e.g. Motion, or DML) must be
+"materialized" into physical tuples.
+
+MemTuple
+
+In Greenplum, MemTuple replaces the MinimalTuple.  MemTuple is also
+serialized tuple format.  The main design goal is to reduce CPU usage
```
Comment (kavinderd) [inline src/backend/executor/README.tuples]: is also *a* serialized tuple format

**Comment by kavinderd** [comment]
@asimrp This is good stuff and all accurate as far as I know. I think there are a few sections that would be very helpful to have:

- Toasted attributes (maybe just a link to the Postgres docs)
- How execution decides which form of a tuple is used

The latter is something we (QP) can add to the doc

File: src/backend/executor/README.tuples [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+What is a tuple representation?
+-------------------------------
+
+Tuples are passed around in generic TupleTableSlot objects. Interested
+parties need to be able to decode the data and see the values. The
+current tuple representation provides information on how to decode the
+data into typed attributes
+
+Why do we need more than one tuple representations?
```
Comment (hlinnaka) [inline src/backend/executor/README.tuples]: s/representations/representation/

File: src/backend/executor/README.tuples (hunk 6 above)
Comment (asimrp) [reply src/backend/executor/README.tuples]: I rewrote it as

'Where a heap tuple is stored is important for resource management as shared buffers need to be correctly pinned'

the reason being scope of a pin includes a whole buffer, it is not granular enough to distinguish tuples within a buffer.

**Comment by asimrp** [comment]
Thank you for the detailed review, @kavinderd.  Pushed to master with comments incorporated.

**Pull request closed without merging**